asterius/rts/rts.float.mjs

```diff
@@ -15,86 +15,229 @@ export class FloatCBits {
+
+    Object.seal(this);
+  }
+
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is
+  isFloatNegativeZero(x) {
+    return Object.is(-0, x);
+  }
+
+  isFloatNaN(x) {
+    return x != x;
+  }
+
+  isDoubleNaN(x) {
+    return x != x;
+  }
+
+  // Remember, floats have 3 states: {finite, infinite, NaN}.
+  isFloatInfinite(x) {
+    return !isFinite(x) && !this.isFloatNaN(x);
+  }
+
+  isDoubleInfinite(x) {
+    return !isFinite(x) && !this.isDoubleNaN(x);
+  }
+
+  // extract the mantissa from the little endian representation of the bits
+  // of the float.
+  // little endian: <0A 0B> stored as mem[p] = 0A, mem[p + 1] = OB
+  floatMantissaFromBits(bits) {
+    const mask = (1 << 23) - 1;
+    return bits & mask;
+  }
+
+  // extract the exponent from the little endian representation of the bits
+  // of the float.
+  floatExponentFromBits(bits) {
+    const mask = (1 << 8) - 1;
+    const sign = this.floatSignFromBits(bits);
+    return ((bits >> 23) & mask) ^ (sign << 31);
+  }
+
+  floatSignFromBits(bits) {
+    return bits >> 31;
+  }
+
+  doubleMantissaFromBits(bits) {
+    const mask = (BigInt(1) << BigInt(52)) - BigInt(1);
+    return bits & mask;
+  }
+
+  doubleExponentFromBits(bits) {
+    const mask = BigInt((1 << 11) - 1);
+    const sign = this.doubleSignFromBits(bits);
+
+    const bitsNoSign = bits ^ (sign << BigInt(63));
+    return (bitsNoSign >> BigInt(52)) & mask;
+  }
+
+  doubleSignFromBits(bits) {
+    return bits >> BigInt(63);
+  }
+
+  // Check if a double is denormal.
+  isDoubleDenormalized(x) {
+    const bits = this.DoubleToIEEE(x);
+
+    console.log("x: %f | bits: %s", x, bits.toString(16));
+    const exponent = this.doubleExponentFromBits(bits);
+    console.log("x: %f | exponent: %s", x, exponent.toString(16));
+    const mantissa = this.doubleMantissaFromBits(bits);
+    console.log("x: %f | mantissa: %s", x, mantissa.toString(16));
+    return exponent === BigInt(0) && mantissa !== BigInt(0);
+  }
+
+  isFloatDenormalized(x) {
+    const bits = this.FloatToIEEE(x);
+    const exponent = this.floatExponentFromBits(bits);
+    const mantissa = this.floatMantissaFromBits(bits);
+    return exponent === 0 && mantissa !== 0;
+  }
+
+  // Does it really make sense to have two functions?  probably not...
+  isDoubleNegativeZero(x) {
+    return Object.is(-0, x);
+  }
+
+  FloatToIEEE(f) {
+    this.view.setFloat32(0, f);
+    return this.view.getUint32(0);
+  }
+
+  DoubleToIEEE(d) {
+    this.view.setFloat64(0, d);
+    return this.view.getBigUint64(0);
+  }
+
+  // return two 32-bit integers, [low, high] from a 64 bit double;
+  DoubleTo2Int(d) {
+    this.view.setFloat64(0, d);
+    const low = this.view.getUint32(0);
+    const high = this.view.getUint32(/*offset=*/ 4);
+    return [low, high];
+  }
+
+  IEEEToFloat(ieee) {
+    this.view.setUint32(0, ieee);
+    return this.view.getFloat32(0);
+  }
+
+  IEEEToDouble(ieee) {
+    this.view.setUint64(0, ieee);
```

Comment:
`DataView.prototype.setBigUint64` actually. And don't forget to handle the `i64`/`f64` coercion for `ieee`.

Reply:
Actually it seems `IEEEToDouble` isn't used anywhere. Is it used in subsequent PRs? If not we can pull some weed here.